src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/SelectExpandNode.cs

```diff
@@ -184,19 +206,103 @@ public SelectExpandNode(SelectExpandClause selectExpandClause, IEdmStructuredTyp
 
         private void BuildExpansions(SelectExpandClause selectExpandClause, HashSet<IEdmNavigationProperty> allNavigationProperties)
         {
-            foreach (SelectItem selectItem in selectExpandClause.SelectedItems)
+            if (selectExpandClause != null)
+            {
+                foreach (SelectItem selectItem in selectExpandClause.SelectedItems)
+                {
+                    ExpandedNavigationSelectItem expandItem = selectItem as ExpandedNavigationSelectItem;
+                    if (expandItem != null)
+                    {
+                        ValidatePathIsSupported(expandItem.PathToNavigationProperty);
+                        NavigationPropertySegment navigationSegment = (NavigationPropertySegment)expandItem.PathToNavigationProperty.LastSegment;
+                        IEdmNavigationProperty navigationProperty = navigationSegment.NavigationProperty;
+                        if (allNavigationProperties.Contains(navigationProperty))
+                        {
+                            ExpandedNavigationProperties.Add(navigationProperty, expandItem.SelectAndExpand);
+                        }
+                        else
+                        {
+                            //This is the case where the navigation property is not on the top-level resource of the request.
+                            int count = expandItem.PathToNavigationProperty.Count;
+                            IEdmProperty complexProperty = null;
+                            for (int i = 0; i < count; i++)
+                            {
+                                ODataPathSegment segment = expandItem.PathToNavigationProperty.ElementAt(i);
+                                PropertySegment propertySegment = segment as PropertySegment;
+                                if (propertySegment != null)
+                                {
+                                    complexProperty = propertySegment.Property;
+                                    break;
+                                }
+                            }
+                            if (complexProperty != null && ExpandedNavigationPropertiesOnComplexTypes.ContainsKey(complexProperty))
```

Comment:
>if (complexProperty != null && ExpandedNavigationPropertiesOnComplexTypes.ContainsKey(complexProperty)) [](start = 28, length = 103)

Do we support this:

~/Customers(1)/Address?$expand=NS.UsAddress/NavigationProperty

Reply:
Yes, we support this. 

---
In reply to: [274694855](https://github.com/OData/WebApi/pull/1738#discussion_r274694855) [](ancestors = 274694855)